hazelcast/src/main/java/com/hazelcast/spi/impl/InternalCompletableFuture.java

```diff
@@ -16,28 +16,142 @@
 
 package com.hazelcast.spi.impl;
 
-import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.internal.serialization.SerializationService;
+import com.hazelcast.nio.serialization.Data;
+
+import javax.annotation.Nonnull;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+import java.util.function.BiConsumer;
 
 /**
- * A {@link com.hazelcast.core.ICompletableFuture} with more functionality like getting
- * the result without needing to deal with checked exceptions.
+ * An extension to {@link CompletableFuture} supporting a {@code joinInternal()}
+ * variant with more relaxed exception throwing conventions.
+ *
+ * This is the base class for all internally used {@link CompletableFuture}s.
+ *
+ * Does not perform optional deserialization. Uses {@link ForkJoinPool#commonPool()}
+ * or thread-per-task executor as default async executor, as {@link CompletableFuture} does.
  *
- * @param <E>
+ * Provides static factory methods for more specific implementations supporting custom
+ * default async executor or deserialization of completion value.
  */
-public interface InternalCompletableFuture<E> extends ICompletableFuture<E> {
+public class InternalCompletableFuture<V> extends CompletableFuture<V> {
+
+    /**
+     * Similarly to {@link #join()}, returns the value when complete or throws an unchecked exception if
+     * completed exceptionally. Unlike {@link #join()}, checked exceptions are not wrapped in {@link CompletionException};
+     * rather they are wrapped in {@link com.hazelcast.core.HazelcastException}s.
+     *
+     * @return the result
+     */
+    public V joinInternal() {
+        try {
+            return join();
+        } catch (CompletionException e) {
+            Throwable cause = e.getCause();
+            throw AbstractInvocationFuture.wrapOrPeel(cause);
+        }
+    }
+
+    /**
+     * Creates a new {@code InternalCompletableFuture} that is already completed with the value given
+     * as argument as its completion value.
+     *
+     * @param result    completion value
+     * @return          a completed future with the given {@code result} as its completion value.
+     */
+    public static InternalCompletableFuture newCompletedFuture(Object result) {
```

Comment:
WDYT about adding a type parameter to methods like this (`public static <V> InternalCompletableFuture<V> ...`), just to lower the amount of "Unchecked assignment" warnings in the code base.

Reply:
Good point, fixed. I kept the `result` parameter as `Object` (instead of `V`) to avoid issues with near-cache code that deals with either `<Data>` or `<V>`. The warnings are now contained within these few static methods instead of being spread out across the codebase.